iTop 3.0.0 badge dashlet : make whole badge clickable

piRGoif opened the pull request:
**Before :** only the text link was clickable (object name, "create" link)

**Now :** the whole badge is clickable.
* When clicking on the "create" label (the corresponding A tag actually) the create form opens.
* Otherwise the list appears.

piRGoif committed:
:lipstick: Add cursor to dashlet-badge

piRGoif committed:
Add JS to handle click anywhere in the badge

Molkobain reviewed (changes requested):
This would bring a way better UX to the badges indeed! ✌

Just a few remark regarding the current implementation to better match the UIBlock concepts and guidelines:
1. All the JS should be in a dedicated jQuery UI widget instantiated in the `layout.js.twig` of the block, not in a separated file, otherwise badges won't work on others pages (Friendly reminder: Blocks must be atomic and independant, meaning that they must work in any page)
2. Use the `data-role` attributes for JS selectors to keep segregation between style and interaction (you can check examples on other blocks)
3. Remove the HTML tags from the selectors (eg. `div`), again for the sake of the segregation as the markup could evolve.
4. Note: The jQuery UI widget provides an easier way to interact with elements within the block, making selectors simpler/clearer. (see points 2. and 3.)

I admit that I'm not fond of using all this JS just to work around the limitation of having an anchor within an anchor; but in this particular case I think it's worth it (and we don't have another choice 😁)

For the record, another (lighter?) implementation could be to have to "main" (list) anchor around the whole badge, than a `<span href="xxx" onclick="Someting to redirect here;">` instead of the anchor but:
- It will not be great for accessibility as the span could be focused
- It would be a shady hack not as easy to maintain as something following the blocks guidelines

piRGoif commented on application/dashboard.class.inc.php:
```diff
@@ -567,6 +567,21 @@ public function Render($oPage, $bEditMode = false, $aExtraParams = array(), $bCa
+  }
+
+	let $listLink = $eventTarget
+    .closest("div.ibo-dashlet-badge")
```
Using `$(this)`should allow to remove this DOM browsing !
TODO tomorrow !

piRGoif replied:
Outdated comment, closing conversation !

piRGoif commented:
We'll need to have a chat about your propositions so I get a clear picture :)

piRGoif commented:
I forgot to say : currently there is just one click handler set in the dashboard code. I would prefer to keep it that way, as having one handler per page is far better than an handler per badge dashlet !

But of course this code is of the DashletBadge component responsibility... So we might add a method in DashletContainer to be implemented in each dashlet components, and having the dashboard code to call this method on each dashlet impl to get the JS code to return ?

piRGoif commented:
Discussed with @Molkobain : having only one handler on the dashboard level is nice, but we need to keep code responsability which is badge dashlet component.

So we agreed to : 
* create a dashlet badge JQuery widget
* instantiate this widget in all badge instance using a JS twig file

We will later on (maybe 3.1.0 ?) improve the API by adding a mean for components to add "static" JS code (added only once event if multiple instances provided). Combodo internal ref N°3710

piRGoif committed:
:rewind: Rollback previous modifications :/

piRGoif committed:
:recycle: New JQuery widget for dashlet badge component

Would be better to have only one handler on the dashboard side but we want to respect code responsibility !
So until we have a system for components to print "static" code (that will be printed only once whatever number of component instances we have in the page), we are doing it the "old" way : calling a JQuery widget in the component JS Twig.

piRGoif committed:
:art: Fix no click in dashlet margins
Now use root dashlet container (we have 2 of them ?!??)

piRGoif commented:
I pushed the last version that use : 
* a JQuery widget, instanciate in each component (JS twig)
* data-role attributes as selectors

Note that a badge dashlet contains 2 containers : one to contain the whole dashlet, and one again created inside the component (?!??)
So to get to the highest container I had to do weird stuffs :/ (see 209b91bd)

Molkobain commented on js/components/dashlet-badge.js:
```diff
@@ -0,0 +1,69 @@
+
+;
+$(function () {
+	$.widget('itop.dashletBadge',
```
Convention is (see other blocks)
```suggestion
	$.widget('itop.dashlet_badge',
```

piRGoif replied:
Will be solved by future commit

Molkobain commented on js/components/dashlet-badge.js:
```diff
@@ -0,0 +1,69 @@
+			options: {},
+			js_selectors:
+				{
+					dashletContainer: '[data-role="ibo-dashlet"]',
+					dashletActions: '[data-role="ibo-dashlet-badge--actions"]',
+					dashletActionList: '[data-role="ibo-dashlet-badge--action-list"]'
+				},
+			dashletContainer: null,
```
No camel case properties, only snake case, see other blocks

piRGoif replied:
Will be solved by future commit

Molkobain commented on sources/application/UI/Base/Component/Dashlet/DashletBadge.php:
```diff
@@ -12,6 +12,10 @@ class DashletBadge extends DashletContainer
 {
 	public const BLOCK_CODE = 'ibo-dashlet-badge';
 	public const DEFAULT_HTML_TEMPLATE_REL_PATH = 'base/components/dashlet/dashletbadge';
+	public const DEFAULT_JS_TEMPLATE_REL_PATH = 'base/components/dashlet/dashletbadge';
```
Template should be `dashlet-badge` (even though I understand you followed the naming from the HTML template, which should be fixed as well; if you could fix it while you are at it, it would be great 👍)

piRGoif replied:
Will be solved by future commit

Molkobain commented on sources/application/UI/Base/Component/Dashlet/DashletBadge.php:
```diff
@@ -12,6 +12,10 @@ class DashletBadge extends DashletContainer
 	public const DEFAULT_HTML_TEMPLATE_REL_PATH = 'base/components/dashlet/dashletbadge';
+	public const DEFAULT_JS_TEMPLATE_REL_PATH = 'base/components/dashlet/dashletbadge';
+	public const DEFAULT_JS_FILES_REL_PATH = [
+		'js/components/dashlet-badge.js',
```
Could you put this file in a `dashlet` sub-folder just like for the templates / stylesheets?

piRGoif replied:
You'll have to explain me directly, not sure to understand ?

piRGoif replied:
Discussed with Guillaume, changed in 59cd940b

Molkobain commented on sources/application/UI/Base/Component/Dashlet/DashletContainer.php:
```diff
@@ -14,4 +14,11 @@ class DashletContainer extends UIContentBlock
+	{
+		parent::__construct($sId, $aContainerClasses);
+
+		$this->AddDataAttribute('role', 'ibo-dashlet');
```
```suggestion
		$this->AddDataAttribute('role', static::BLOCK_CODE);
```

piRGoif replied:
Will be solved by future commit

Molkobain commented on sources/application/UI/Base/Component/Dashlet/DashletContainer.php:
```diff
@@ -14,4 +14,11 @@ class DashletContainer extends UIContentBlock
 	public const BLOCK_CODE = 'ibo-dashlet';
 	public const DEFAULT_HTML_TEMPLATE_REL_PATH = 'base/layouts/content-block/layout';
+
+	public function __construct(string $sId = null, array $aContainerClasses = [])
```
I'm not fond of this at it's clear that there is clearly a design problem in the dashlets block, this should be in a parent template, but for now I might it's acceptable. We might have to refactor the dashlets a bit later 😥

piRGoif replied:
Comment well noted I agree as said in our IM !
Out of scope for now, so resolving this conversation

piRGoif committed:
:truck: Rename JQuery widget

piRGoif committed:
:truck: Var names changed for snake case

piRGoif committed:
DashletContainer : use constant for data-role instead of hardcoded string

piRGoif committed:
:truck: Rename dashlet badge templates

piRGoif commented:
Changing state : not a draft anymore

piRGoif committed:
:truck: Move dashlet-badge.js

piRGoif commented:
@Molkobain is OK to merge, so let's go !
